src/ucp/core/ucp_worker.c

```diff
@@ -507,6 +508,60 @@ ucs_status_t ucp_worker_set_ep_failed(ucp_worker_h worker, ucp_ep_h ucp_ep,
     return ret_status;
 }
 
+static unsigned ucp_worker_set_ep_failed_progress(void *arg)
+{
+    ucp_worker_set_ep_failed_arg_t *set_ep_failed_arg = arg;
+    ucp_worker_h worker                               =
+            set_ep_failed_arg->ucp_ep->worker;
+
+    UCS_ASYNC_BLOCK(&worker->async);
+    ucp_worker_set_ep_failed(worker, set_ep_failed_arg->ucp_ep,
+                             set_ep_failed_arg->lane,
+                             set_ep_failed_arg->status);
+    UCS_ASYNC_UNBLOCK(&worker->async);
+
+    ucs_free(set_ep_failed_arg);
+    return 1;
+}
+
+int ucp_worker_set_ep_failed_remove_filter(const ucs_callbackq_elem_t *elem,
+                                           void *arg)
+{
+    ucp_worker_set_ep_failed_arg_t *set_ep_failed_arg = arg;
+
+    if ((elem->cb == ucp_worker_set_ep_failed_progress) &&
+        (set_ep_failed_arg->ucp_ep == arg)) {
+        ucs_free(set_ep_failed_arg);
+        return 1;
+    }
+
+    return 0;
+}
+
+void ucp_worker_set_ep_failed_schedule(ucp_worker_h worker, ucp_ep_h ucp_ep,
```

Comment:
- no need to pass worker
- rename to ucp_ep_set_failed_safe/ucp_ep_set_failed_schedule/ucp_ep_set_failed_invoke/...
- move these funcs to ucp_ep.c? since they operate on a particular ep

Reply:
done